### PR TITLE
Attempt to change API endpoint per #83

### PR DESCRIPTION
**Pull request by DeathCamel58**
Please test this a bit more than I did. It seems to be working for me with postman and such, but I can't be 100% sure, because I'm not 100% sure what to test other than getting co-ordinates from a zip code, then finding meetings nearby.
Issue #83

### CODE REVIEW

**Comment by radius314** [comment]
Could you rebase this and open against unstable?  Master is frozen right now until unstable gets merged in.

**Comment by jbraswell** [comment]
I really like this. I think that if a google maps api key is configured, it should use google maps, but that the default should be OSM so that yap just works "out of the box".

**Comment by pjaudiomv** [comment]
@jbraswell I agree with this i think google api should still be an option. the only thing im not sure is how to get the timezonedata im not sure of any free apis that you dont need to register and get a key for. early versions of yap used timezonedb but that was with a key danny had created

File: functions.php [hunk 1]
```diff
@@ -51,7 +51,7 @@ function getCoordinatesForAddress($address) {
 
     if (strlen($address) > 0) {
         $map_details_response = get($GLOBALS['maps_endpoint']
-            . "&address="
+            . "&postalcode="
```
Comment (pjaudiomv) [inline functions.php]: this should probably be "&q=" as we accept both address and postalcode data

File: functions.php [hunk 2]
```diff
@@ -50,16 +50,15 @@ function getCoordinatesForAddress($address) {
     $coordinates = new Coordinates();
 
     if (strlen($address) > 0) {
-        $map_details_response = get($GLOBALS['google_maps_endpoint']
+        $map_details_response = get($GLOBALS['maps_endpoint']
             . "&address="
             . urlencode($address)
-            . (isset($GLOBALS['location_lookup_bias']) ? "&components=" . urlencode($GLOBALS['location_lookup_bias']) : ""));
+            . (isset($GLOBALS['location_lookup_bias']) ? "&country=" . urlencode($GLOBALS['location_lookup_bias']) : ""));
```
Comment (pjaudiomv) [inline functions.php]: this would be better if it was "&countrycodes=" as this would force the results to be a specific country better

**Comment by jbraswell** [comment]
@pjaudiomv It looks like this PR is setting a new `$timezone_lookup_endpoint`. Is there something else that I'm missing? You and Danny are way more familiar with this codebase than I am.

```
$timezone_lookup_endpoint = "https://nominatim.openstreetmap.org/search?format=json&limit=1&addressdetails=1"; 
```

**Comment by pjaudiomv** [comment]
that endpoint doesnt contain timezone data and to my knowledge openstreetmaps doesnt give you timezone data in any api. (I hope it does, but i dont think so) also most timezone lookup apis do so based off long/lats. google, azure, geonames, timezonedb etc

**Comment by jbraswell** [comment]
geonames.org seems to be a viable alternative for time zone data. It does require user registration, but it’s quick and free. Way easier than setting up a google api key.

**Comment by jbraswell** [comment]
Example: http://api.geonames.org/timezoneJSON?formatted=true&lat=47.01&lng=10.2&username=demo&style=full

**Comment by pjaudiomv** [comment]
Nice! Yea timezonedb is easy to setup to, and I think there database is more acurate.

**Comment by DeathCamel58** [comment]
Hey @radius314 and @jbraswell, this edit has yap pull from OSM's Nominatim API by default. When a Google Maps API key is set in `config.php`, yap will stop pulling from Nominatim, and will instead try Google Maps. 
I haven't implemented an alternative for the time zone api lookup yet (I will soon), but even without an API key, Google will return the time zone information for a certain number of requests (I think it's a few thousand per day)

File: functions.php [hunk 3]
```diff
@@ -242,15 +244,29 @@ function getCoordinatesForAddress($address) {
...
+            $map_details = json_decode($map_details_response);
+            if (count($map_details->results) > 0) {
+                $coordinates->location  = $map_details->results[0]->formatted_address;
+                $geometry               = $map_details->results[0]->geometry->location;
+                $coordinates->latitude  = $geometry->lat;
+                $coordinates->longitude = $geometry->lng;
+            }
+        } else {
+            $map_details_response = get($GLOBALS['osm_maps_endpoint']
+                . "&postalcode="
```
Comment (pjaudiomv) [inline functions.php]: this should probably be "&q=" as address lookup queries aren't limited to postalcodes

File: functions.php [hunk 4]
```diff
@@ -242,15 +244,29 @@ function getCoordinatesForAddress($address) {
...
+                . urlencode($address)
+                . (isset($GLOBALS['location_lookup_bias']) ? "&country=" . urlencode($GLOBALS['location_lookup_bias']) : ""));
```
Comment (pjaudiomv) [inline functions.php]: this should probably be "&countrycodes=" using the ISO 3166-1alpha2 code, per nominatim spec it will honor this, but will not always with just &country

**Comment by pjaudiomv** [comment]
the parameters in which the location_lookup_bias is set are different between google and osm, maybe we default to only do location lookup bias for google to be by country (components=country:) so they can be the same (both defaults would be US) idk

**Comment by pjaudiomv** [comment]
it would also be sweet to not only check whether the $google_maps_api_key is set but that if it fails the upgrade-advisor checks as well to then default to OSM

File: functions.php [hunk 5]
```diff
@@ -32,6 +32,8 @@
 
 $google_maps_endpoint = "https://maps.googleapis.com/maps/api/geocode/json?key=" . trim($google_maps_api_key);
 $timezone_lookup_endpoint = "https://maps.googleapis.com/maps/api/timezone/json?key" . trim($google_maps_api_key);
+$osm_maps_endpoint = "https://nominatim.openstreetmap.org/search?format=json&limit=1&addressdetails=1";  
+$osm_timezone_lookup_endpoint = "https://nominatim.openstreetmap.org/search?format=json&limit=1&addressdetails=1";
```
Comment (radius314) [inline functions.php]: could you escapsulate the selection of the endpoint into its own function?  this makes it re-usable without having to litter application logic with the endpoint selection.

Comment (radius314) [reply functions.php]: let me rephrase this, write a generic function that calls out either maps or timezone endpoints and maps to consistent response values.  decouple from application logic.

**Comment by DeathCamel58** [comment]
I think this is ready to roll, I just patched an issue that this caused (it always searched nominatm, bad variable check). Any input anyone?

**Comment by radius314** [comment]
was the timezone issue sorted out?

**Comment by DeathCamel58** [comment]
@radius314, the timezone lookups work, it's just limited to 1 time zone lookup per second. I totally forgot about that btw, I'll implement something tonight.
